File: aviquest/lib/models/bird.dart
import 'package:flutter/material.dart';

const rarityColors = {
  'common': Colors.white70,
  'uncommon': Color(0xFF4CAF50),
  'rare': Color(0xFF2196F3),
  'legendary': Colors.amber,
  'unknown': Color(0xFFCE93D8),
};

class Bird {
  final String name;
  final String scientificName;
  final String imageUrl;
  final String audioUrl;
  final String lore;
  final String habitat;
  final String conservationStatus;
  final String rarity;
  final int baseXp;

  const Bird({
    required this.name,
    required this.scientificName,
    required this.imageUrl,
    required this.audioUrl,
    required this.lore,
    required this.habitat,
    required this.conservationStatus,
    required this.rarity,
    required this.baseXp,
  });

  factory Bird.fromJson(Map<String, dynamic> json) {
    return Bird(
      name: json['name'] as String,
      scientificName: json['scientificName'] as String,
      imageUrl: json['imageUrl'] as String,
      audioUrl: json['audioUrl'] as String,
      lore: json['lore'] as String,
      habitat: json['habitat'] as String,
      conservationStatus: json['conservationStatus'] as String,
      rarity: json['rarity'] as String,
      baseXp: json['baseXp'] as int,
    );
  }

  Color get rarityColor => rarityColors[rarity] ?? Colors.white70;

  int get xp {
    switch (rarity) {
      case 'uncommon':
        return (baseXp * 1.5).round();
      case 'rare':
        return baseXp * 2;
      case 'legendary':
        return baseXp * 5;
      default:
        return baseXp;
    }
  }
}

Bird unknownBird(String name) => Bird(
      name: name,
      scientificName: 'Species not yet in database',
      imageUrl: '',
      audioUrl: '',
      lore:
          'You found something we\'ve never seen before! This species isn\'t in our database yet. '
          'Your discovery has been logged and will help us grow AviQuest.',
      habitat: 'Unknown',
      conservationStatus: 'Unknown',
      rarity: 'unknown',
      baseXp: 100,
    );

File: temporal_backend/models/bird.py
from __future__ import annotations

import enum
from dataclasses import dataclass


class Rarity(str, enum.Enum):
    COMMON = "common"
    UNCOMMON = "uncommon"
    RARE = "rare"
    LEGENDARY = "legendary"


# XP multipliers per rarity tier (mirrors Flutter client logic)
RARITY_XP: dict[Rarity, int] = {
    Rarity.COMMON: 50,
    Rarity.UNCOMMON: 120,
    Rarity.RARE: 300,
    Rarity.LEGENDARY: 750,
}


@dataclass
class Bird:
    name: str
    scientific_name: str
    image_url: str
    audio_url: str
    habitat: str
    conservation_status: str
    rarity: Rarity
    base_xp: int

    @property
    def xp_value(self) -> int:
        return self.base_xp or RARITY_XP.get(self.rarity, 50)
